numba/tests/test_array_manipulation.py

```diff
@@ -754,6 +762,74 @@ def _assert_raises(arr, val):
+            [np.ones(1), (1, 2, 3)],
+            [np.arange(3), (3,)],
+            [np.arange(3), (1, 3)],
+            [np.arange(3), (2, 3)],
+            # test if shape is not a tuple
+            [np.ones(0), 0],
+            [np.ones(1), 1],
+            [np.ones(1), 2],
+            # these cases with size 0 are strange, but they reproduce the behavior
+            # of broadcasting with ufuncs (see test_same_as_ufunc above)
+            [np.ones(1), (0,)],
+            [np.ones((1, 2)), (0, 2)],
+            [np.ones((2, 1)), (2, 0)],
+        ]
+        for input_array, shape in data:
+            expected = pyfunc(input_array, shape)
+            got = cfunc(input_array, shape)
+            self.assertPreciseEqual(got, expected)
+
+    def test_broadcast_to_raises(self):
+        pyfunc = numpy_broadcast_to
+        cfunc = jit(nopython=True)(pyfunc)
+
+        # Tests taken from
+        # https://github.com/numpy/numpy/blob/75f852edf94a7293e7982ad516bee314d7187c2d/numpy/lib/tests/test_stride_tricks.py#L260-L276  # noqa: E501
+        data = [
+            [(0,), (), TypingError, 'The argument "shape" must be a tuple or an integer.'],
+            [(1,), (), TypingError, 'The argument "shape" must be a tuple or an integer.'],
+            [(3,), (), TypingError, 'The argument "shape" must be a tuple or an integer.'],
+            [(3,), (1,), ValueError, 'operands could not be broadcast together with remapped shapes'],
+            [(3,), (2,), ValueError, 'operands could not be broadcast together with remapped shapes'],
+            [(3,), (4,), ValueError, 'operands could not be broadcast together with remapped shapes'],
+            [(1, 2), (2, 1), ValueError, 'operands could not be broadcast together with remapped shapes'],
+            [(1, 1), (1,), ValueError, 'input operand has more dimensions than allowed by the axis remapping'],
+            [(2, 2), (3,), ValueError, 'input operand has more dimensions than allowed by the axis remapping'],
+            [(1,), -1, ValueError, 'all elements of broadcast shape must be non-negative'],
+            [(1,), (-1,), ValueError, 'all elements of broadcast shape must be non-negative'],
+            [(1, 2), (-1, 2), ValueError, 'all elements of broadcast shape must be non-negative'],
+        ]
+        for orig_shape, target_shape, err, msg in data:
+            self.disable_leak_check()
+            arr = np.zeros(orig_shape)
+            with self.assertRaises(err) as raises:
+                cfunc(arr, target_shape)
+            self.assertIn(msg, str(raises.exception))
+
+    def test_broadcast_to_indexing(self):
+        pyfunc = numpy_broadcast_to_indexing
+        cfunc = jit(nopython=True)(pyfunc)
+        data = [
+            [np.ones(2), (2, 2), (1,)],
+        ]
+        for input_array, shape, idx in data:
+            expected = pyfunc(input_array, shape, idx)
+            got = cfunc(input_array, shape, idx)
+            self.assertPreciseEqual(got, expected)
```

Comment:
Should there be some testing of the "view" nature of the broadcasted array, i.e. that mutating the original mutates the broadcast? Examples:
```python
from numba import njit
import numpy as np

@njit
def foo(arr, shape):
    return np.broadcast_to(arr, shape)

x = np.ones(2)
ret = foo(x, (6, 2))
print(ret)

x[0] = 10
print(ret)
```
and
```python
from numba import njit
import numpy as np

@njit
def foo(arr, shape):
    tmp = np.broadcast_to(arr, shape)
    print(tmp)
    arr[0] = 10
    print(tmp)

x = np.ones(2)
ret = foo(x, (6, 2))
```